desktopexporter/internal/store/util.go

```diff
@@ -0,0 +1,243 @@
+)
+
+// Helper function to convert Events to DuckDB list of STRUCT string format
+func eventToString(events []telemetry.EventData) string {
+	eventStrings := []string{}
+
+	for _, event := range events {
+		attributes := mapToString(event.Attributes)
+		eventStrings = append(eventStrings, fmt.Sprintf("{name: '%s', timestamp: '%v', attributes: %s, droppedAttributesCount: %d}",
+			escapeString(event.Name),
+			event.Timestamp.Format(time.RFC3339Nano),
+			attributes,
+			event.DroppedAttributesCount))
+	}
+	return fmt.Sprintf("[%s]", strings.Join(eventStrings, ", "))
+}
+
+// Helper function to convert Links to DuckDB list of STRUCT string format
+func linkToString(links []telemetry.LinkData) string {
+	linkStrings := []string{}
+
+	for _, link := range links {
+		attributes := mapToString(link.Attributes)
+		linkStrings = append(linkStrings, fmt.Sprintf(
+			"{traceID: '%s', spanID: '%s', traceState: '%s', attributes: %s, droppedAttributesCount: %d}",
+			escapeString(link.TraceID),
+			escapeString(link.SpanID),
+			escapeString(link.TraceState),
+			attributes,
+			link.DroppedAttributesCount))
+	}
+	return fmt.Sprintf("[%s]", strings.Join(linkStrings, ", "))
+}
+
+// Helper function to convert map to DuckDB MAP string format
+func mapToString(m map[string]interface{}) string {
+	var pairs []string
+	for k, v := range m {
+		var valStr string
+		switch v := v.(type) {
+		case string:
+			valStr = fmt.Sprintf("'%s'::attribute", escapeString(v))
+		case int, int32, int64:
+			valStr = fmt.Sprintf("%d::attribute", v)
+		case float32, float64:
+			valStr = fmt.Sprintf("%f::attribute", v)
+		case bool:
+			if v {
+				valStr = "true::attribute"
+			} else {
+				valStr = "false::attribute"
+			}
+		case []string:
+			elements := make([]string, len(v))
+			for i, s := range v {
+				elements[i] = fmt.Sprintf("'%s'::attribute", escapeString(s))
+			}
+			valStr = fmt.Sprintf("[%s]", strings.Join(elements, ", "))
+		case []int64:
+			elements := make([]string, len(v))
+			for i, n := range v {
+				elements[i] = fmt.Sprintf("%d::attribute", n)
+			}
+			valStr = fmt.Sprintf("[%s]", strings.Join(elements, ", "))
+		case []float64:
+			elements := make([]string, len(v))
+			for i, f := range v {
+				elements[i] = fmt.Sprintf("%f::attribute", f)
+			}
+			valStr = fmt.Sprintf("[%s]", strings.Join(elements, ", "))
+		case []bool:
+			elements := make([]string, len(v))
+			for i, b := range v {
+				if b {
+					elements[i] = "true::attribute"
+				} else {
+					elements[i] = "false::attribute"
+				}
+			}
+			valStr = fmt.Sprintf("[%s]", strings.Join(elements, ", "))
+		default:
+			valStr = fmt.Sprintf("union_value(str := '%v')", v)
+		}
+		pairs = append(pairs, fmt.Sprintf("'%s': %v", escapeString(k), valStr))
+	}
+	return fmt.Sprintf("MAP{%s}", strings.Join(pairs, ", "))
+}
+
+// Helper function to parse events from raw DuckDB [STRUCT(...)] string format
+func parseRawEvents(rawEvents string) []telemetry.EventData {
```

Comment:
Returning empty values on error is a reasonable choice here, but we should probably at least log that this happened. Tests would also help document this behavior.